src/main/frontend/db/query_dsl.cljs

```diff
@@ -324,28 +325,18 @@
 
 (defn- add-bindings!
   [q]
-  (let [b? (atom false)
-        p? (atom false)
-        not? (atom false)]
-    (walk/postwalk (fn [f]
-                     (cond
-                       (= 'not f)
-                       (reset! not? true)
-                       (= '?b f)
-                       (reset! b? true)
-                       (= '?p f)
-                       (reset! p? true)
-                       :else
-                       f)) q)
-    (if @not?
+  (let [syms ['?b '?p 'not]
+        [b? p? not?] (-> (set/intersection (set syms) (set (flatten q)))
+                         (map syms))]
```

Comment:
Not sure whether the construction of `(set (flatten q))` is lazy, but if it isn't you can avoid the overhead by filtering first:

```clojure
(let [syms '[?b ?p not]
      [b? p? not?] (-> (keep (set syms) (flatten q))
                       (set)
                       (map syms))]
  ...)
```

Reply:
The `set` should be eager (loop, recur and persistent!)